system-addon/common/Dedupe.jsm

```diff
@@ -0,0 +1,32 @@
+this.Dedupe = class Dedupe {
+  constructor(createKey, compare) {
+    this.createKey = createKey || this.defaultCreateKey;
+    this.compare = compare || this.defaultCompare;
+  }
+
+  defaultCreateKey(item) {
+    return item;
+  }
+
+  defaultCompare() {
+    return false;
+  }
+
+  /**
+   * Removes duplicates in a list based on createKey.
```

Comment:
This should probably say it picks the first item that it finds. So re: @k88hudson's question about pinned, it will remove the pinned if a non-pinned duplicate key item appeared first.